rotkehlchen/accounting/cost_basis/base.py

```diff
@@ -98,17 +100,74 @@ def __str__(self) -> str:
         )
 
 
-@dataclass(init=True, repr=True, eq=True, order=False, unsafe_hash=False, frozen=False)
+class BaseAcquisitionsOrder(Iterator, metaclass=ABCMeta):
+    _acquisitions: deque[AssetAcquisitionEvent]
+
+    def __init__(self) -> None:
+        self._acquisitions = deque()
+
+    @abstractmethod
+    def add_acquisition(self, acquisition: AssetAcquisitionEvent) -> None:
+        """The core method. Should be implemented by subclasses.
+            This method takes a new acquisition and decides where to insert it
+            This method determines the PnL order
+        """
+        ...
+
+    def __next__(self) -> Optional[AssetAcquisitionEvent]:
+        """Iteration method over acquisition events. Returns current front event
+        May raise:
+            StopIterationError - when we have iterated over all events
+        """
+        try:
+            return self._acquisitions[0]
+        except IndexError as e:
+            raise StopIteration from e
+
+    def report_result(self, used_amount: FVal) -> None:
+        """This function should be used to report results of the
+        currently processed event (received from __next__)
+        The current event's remaining_amount will be decreased by used_amount
+        If event's remaining_amount will become ZERO, the event will be deleted
+        May raise:
+            ValueError if used_amount is incorrect
+            IndexError if the method was called when acquisitions were empty
+        """
+        if used_amount < ZERO or used_amount > self._acquisitions[0].remaining_amount:
+            raise ValueError(f'Used amount should be in the interval [0, {self._acquisitions[0].remaining_amount}]')  # noqa: E501
+
+        self._acquisitions[0].remaining_amount -= used_amount
+        if self._acquisitions[0].remaining_amount == ZERO:
+            self._acquisitions.popleft()
+
+
+class FIFOAcquisitionsOrder(BaseAcquisitionsOrder):
+    """Accounting in FIFO (first-in-first-out) order"""
+    def add_acquisition(self, acquisition: AssetAcquisitionEvent) -> None:
+        self._acquisitions.append(acquisition)
+
+
+class LIFOAcquisitionsOrder(BaseAcquisitionsOrder):
+    """Accounting in LIFO (last-in-first-out) order"""
+    def add_acquisition(self, acquisition: AssetAcquisitionEvent) -> None:
+        self._acquisitions.appendleft(acquisition)
+
+
 class CostBasisEvents:
-    used_acquisitions: List[AssetAcquisitionEvent] = field(init=False)
```

Comment:
Changed it from dataclass to a usual class since it was not used as a data class + we need to pass an argument to the constructor to select accounting order